File: remote/src/main/java/com/kelsos/mbrc/mappers/PlaylistTrackMapper.java
package com.kelsos.mbrc.mappers;

import com.kelsos.mbrc.dao.PlaylistDao;
import com.kelsos.mbrc.dao.PlaylistTrackDao;
import com.kelsos.mbrc.dao.PlaylistTrackInfoDao;
import com.kelsos.mbrc.dao.views.PlaylistTrackView;
import com.kelsos.mbrc.dto.playlist.PlaylistTrack;
import com.kelsos.mbrc.interfaces.ItemProvider;
import java.util.List;
import rx.Observable;

public class PlaylistTrackMapper {
  public static List<PlaylistTrackDao> map(List<PlaylistTrack> data,
      ItemProvider<PlaylistDao> playlistItemProvider,
      ItemProvider<PlaylistTrackInfoDao> infoDaoItemProvider) {
    return Observable.from(data)
        .map(value -> map(value, playlistItemProvider, infoDaoItemProvider))
        .toList().toBlocking().first();
  }

  public static PlaylistTrackDao map(PlaylistTrack data,
      ItemProvider<PlaylistDao> playlistItemProvider,
      ItemProvider<PlaylistTrackInfoDao> infoDaoItemProvider) {

    PlaylistTrackDao dao = new PlaylistTrackDao();
    dao.setId(data.getId());
    dao.setPlaylist(playlistItemProvider.getById(data.getPlaylistId()));
    dao.setTrackInfo(infoDaoItemProvider.getById(data.getTrackInfoId()));
    dao.setPosition(data.getPosition());
    dao.setDateAdded(data.getDateAdded());
    dao.setDateUpdated(data.getDateUpdated());
    dao.setDateDeleted(data.getDateDeleted());

    return dao;
  }

  public static com.kelsos.mbrc.domain.PlaylistTrack map(PlaylistTrackView view) {
    com.kelsos.mbrc.domain.PlaylistTrack track = new com.kelsos.mbrc.domain.PlaylistTrack();
    track.setPath(view.getPath());
    track.setPosition(view.getPosition());
    track.setId(view.getId());
    track.setArtist(view.getArtist());
    track.setTitle(view.getTitle());
    return track;
  }
}

File: mbrc-lib/src/main/kotlin/com/kelsos/mbrc/mappers/PlaylistTrackMapper.kt
package com.kelsos.mbrc.mappers

import com.kelsos.mbrc.dao.PlaylistDao
import com.kelsos.mbrc.dao.PlaylistTrackDao
import com.kelsos.mbrc.dao.PlaylistTrackInfoDao
import com.kelsos.mbrc.dao.views.PlaylistTrackView
import com.kelsos.mbrc.dto.playlist.PlaylistTrack
import com.kelsos.mbrc.interfaces.ItemProvider

object PlaylistTrackMapper {
    fun map(data: List<PlaylistTrack>,
            playlistItemProvider: ItemProvider<PlaylistDao>,
            infoDaoItemProvider: ItemProvider<PlaylistTrackInfoDao>): List<PlaylistTrackDao> {
        return data.map {
            map(it, playlistItemProvider, infoDaoItemProvider)
        }.toList()
    }

    fun map(data: PlaylistTrack,
            playlistItemProvider: ItemProvider<PlaylistDao>,
            infoDaoItemProvider: ItemProvider<PlaylistTrackInfoDao>): PlaylistTrackDao {

        val dao = PlaylistTrackDao()
        dao.id = data.id
        dao.playlist = playlistItemProvider.getById(data.playlistId)
        dao.trackInfo = infoDaoItemProvider.getById(data.trackInfoId)
        dao.position = data.position
        dao.dateAdded = data.dateAdded
        dao.dateUpdated = data.dateUpdated
        dao.dateDeleted = data.dateDeleted

        return dao
    }

    fun map(view: PlaylistTrackView): com.kelsos.mbrc.domain.PlaylistTrack {
        val track = com.kelsos.mbrc.domain.PlaylistTrack()
        track.path = view.path
        track.position = view.position
        track.id = view.id
        track.artist = view.artist
        track.title = view.title
        return track
    }
}
